bbq/compiler/desugar.go

```diff
@@ -247,58 +202,286 @@ func (d *Desugar) VisitFunctionDeclaration(declaration *ast.FunctionDeclaration)
 	)
 }
 
-func (d *Desugar) desugarConditions(
+// Assign the return value to the temporary `$_result` synthetic variable.
+func (d *Desugar) tempResultAssignment(
+	originalReturnStmt *ast.ReturnStatement,
+	returnStmtTypes sema.ReturnStatementTypes,
+) *ast.AssignmentStatement {
+
+	transfer := ast.TransferOperationCopy
+	if returnStmtTypes.ReturnType.IsResourceType() {
+		transfer = ast.TransferOperationMove
+	}
+
+	tempResultAssignmentStmt := ast.NewAssignmentStatement(
+		d.memoryGauge,
+		d.tempResultIdentifierExpr(originalReturnStmt.StartPos),
+		ast.NewTransfer(
+			d.memoryGauge,
+			transfer,
+			originalReturnStmt.StartPos,
+		),
+		originalReturnStmt.Expression,
+	)
+
+	d.elaboration.SetAssignmentStatementTypes(
+		tempResultAssignmentStmt,
+		sema.AssignmentStatementTypes{
+			ValueType:  returnStmtTypes.ValueType,
+			TargetType: returnStmtTypes.ReturnType,
+		},
+	)
+	return tempResultAssignmentStmt
+}
+
+// Return the value that was temporary stored in `$_result` synthetic variable.
+func (d *Desugar) tempResultReturnStatement(
+	hasReturn bool,
+	originalReturnStmt *ast.ReturnStatement,
+	returnType sema.Type,
+) *ast.ReturnStatement {
+
+	var returnExpr ast.Expression
+	var astRange ast.Range
+
+	if hasReturn {
+		if originalReturnStmt.Expression != nil {
+			// `return $_result`
+			returnExpr = d.tempResultIdentifierExpr(originalReturnStmt.StartPos)
+
+			// If the return type is resource, then wrap it with a move.
+			if returnType.IsResourceType() {
+				// i.e: `return <- $_result`
+				returnExpr = ast.NewUnaryExpression(
+					d.memoryGauge,
+					ast.OperationMove,
+					returnExpr,
+					originalReturnStmt.StartPos,
+				)
+			}
+
+		}
+		astRange = originalReturnStmt.Range
+	} else {
+		astRange = ast.EmptyRange
+	}
+
+	return ast.NewReturnStatement(
+		d.memoryGauge,
+		returnExpr,
+		astRange,
+	)
+}
+
+// Declare a `$_result` synthetic variable, to temporarily hold return values.
+func (d *Desugar) declareTempResultVariable(
+	declaration *ast.FunctionDeclaration,
+	pos ast.Position,
+	modifiedStatements []ast.Statement,
+) []ast.Statement {
+	tempResultVarDecl := ast.NewVariableDeclaration(
+		d.memoryGauge,
+		ast.AccessNotSpecified,
+		false,
+		ast.NewIdentifier(
+			d.memoryGauge,
+			tempResultVariableName,
+			pos,
+		),
+		declaration.ReturnTypeAnnotation,
+
+		// TODO: What should be the value?
+		//  We just need the variable to be defined. Value is assigned later.
+		nil,
```

Comment:
Here we only declare the temp variable `var $_result: T`. Then each return statement (eventually) will assign the return value to this `$_result` variable. So this is an uninitialized variable at this point. Maybe is there a better way to handle this?